src/framework-ondata.js

```diff
@@ -1,18 +1,129 @@
 /**
  * File: framework-ondata.js 
  */
+var _ = require('underscore');
+var util = require('util');
+
+var stateFactory = require('./create-state');
+var vniManager = require('./vni-manager');
 
 /** 
- * This is a stub file to allow for a smaller PR and merge.  It will be replaced
- * with the real version of the code once the pipeline-component-factory is finalized.
- *
+ * The standard RDF pipeline framework ondata method. This is called whenever  
+ * there is new data on any port.  This method will update the input state,
+ * and determine if the component updater should be executed now or not. 
+ * If the updater should be executed, the fRunUpdater wrapper API will be
+ * called to to handle that.  The updater output (or error) will be stored 
+ * in output or error state, and those results will be sent on to any 
+ * down stream components for their processing.
+ * 
  * @this port context
- *
- * @param payload data payload on the input port
- * @param socketIndex (optional) this will be set only for ports configured for
- *        multiple inputs (multi/noflo addressable).
+ * @param payload data coming into the port
+ * @param socketIndex socket index if the port is a multiple input (addressable) 
  */
 module.exports = function( payload, socketIndex ) {
 
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     // Set the input states
+     vni.inputStates( portName, socketIndex, stateFactory( vnid, payload ) );
+
+     // check if should run updater & run if necessary
+     if ( shouldRunUpdater( this.nodeInstance, vni ) ) { 
+          
+        // Save vars we will need in the promise where the context is different
+        var outputPorts = this.nodeInstance.outPorts;
+
+        if ( _.isUndefined( this.nodeInstance.wrapper.fRunUpdater ) || 
+             ! _.isFunction( this.nodeInstance.wrapper.fRunUpdater ) ) { 
+
+             // Don't have a wrapper fRunUpdater 
+             throw Error( 'No wrapper fRunUpdater function found!  Cannot run updater.' );
+
+        } else { 
+            var promise = this.nodeInstance.wrapper.fRunUpdater.call( vni );
```

Comment:
Would it make sense for the fRunUpdater to be called from here *inside* a promise, so that if it throws an error then it will be caught and the reject path will be taken, instead of requiring every fRunUpdater to catch errors?